// @easy-webpack/config-uglify/dist/index.js
"use strict";
var core_1 = require('@easy-webpack/core');
var webpack = require('webpack');
module.exports = function uglify(_a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.debug, debug = _c === void 0 ? false : _c, _d = _b.exclude, exclude = _d === void 0 ? [] : _d, _e = _b.mangle, mangle = _e === void 0 ? { screw_ie8: true, keep_fnames: true } : _e;
    return function uglify() {
        var options = debug ? {
            beautify: true,
            mangle: false,
            dead_code: false,
            unused: false,
            deadCode: false,
            compress: {
                screw_ie8: true,
                keep_fnames: true,
                drop_debugger: false,
                dead_code: false,
                unused: false
            },
            comments: true,
        } : {
            beautify: false,
            mangle: mangle,
            exclude: exclude,
            compress: {
                screw_ie8: true,
                warnings: false
            },
            comments: false //prod
        };
        return {
            plugins: core_1.get(this, 'plugins', []).concat([
                // NOTE: To debug prod builds uncomment //debug lines and comment //prod lines
                new webpack.optimize.UglifyJsPlugin(options),
            ])
        };
    };
};
