// @easy-webpack/config-babel/dist/index.js
"use strict";
var core_1 = require('@easy-webpack/core');
var path = require('path');
module.exports = function babel(_a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.options, options = _c === void 0 ? {
        "plugins": [
            "transform-decorators-legacy",
            "transform-class-properties"
        ],
        "presets": [
            [
                "env",
                {
                    "targets": {
                        "browsers": [
                            "last 2 versions",
                            "not ie <= 11"
                        ]
                    },
                    "loose": true,
                    "modules": false
                }
            ]
        ],
        cacheDirectory: true,
    } : _c, _d = _b.exclude, exclude = _d === void 0 ? null : _d;
    return function babel() {
        return {
            module: {
                rules: core_1.get(this, 'module.rules', []).concat([{
                        test: /\.jsx?$/,
                        loader: 'babel-loader',
                        exclude: exclude || (this.metadata.root ? [path.join(this.metadata.root, 'node_modules')] : []),
                        query: options
                    }])
            }
        };
    };
};
