// @easy-webpack/config-aurelia/dist/index.js
"use strict";
var core_1 = require('@easy-webpack/core');
var AureliaWebpackPlugin = require('aurelia-webpack-plugin');
module.exports = function aurelia(allOptions) {
    if (allOptions === void 0) { allOptions = { root: '', src: '', title: 'Aurelia', baseUrl: '/' }; }
    if (!allOptions.customViewLoaders) {
        // temporary fix until released: https://github.com/aurelia/webpack-plugin/commit/472558fc33862832b4192896434946fd935e429e
        allOptions.customViewLoaders = {
            '.css': ['css-loader'],
            '.scss': ['css-loader', 'sass-loader'],
            '.sass': ['css-loader', 'sass-loader'],
            '.less': ['css-loader', 'less-loader'],
            '.styl': ['css-loader', 'stylus-loader'],
        };
    }
    var title = allOptions.title, baseUrl = allOptions.baseUrl, root = allOptions.root, src = allOptions.src;
    return function aurelia() {
        return {
            metadata: {
                title: title,
                baseUrl: baseUrl,
                root: root,
                src: src,
            },
            resolve: {
                modules: [src].concat(core_1.get(this, 'resolve.modules', ['node_modules']))
            },
            plugins: [
                new AureliaWebpackPlugin(allOptions)
            ].concat(core_1.get(this, 'plugins', []))
        };
    };
};
