// @easy-webpack/config-generate-index-html/dist/index.js
"use strict";
var core_1 = require('@easy-webpack/core');
var HtmlWebpackPlugin = require('html-webpack-plugin');
module.exports = function generateIndexHtml(_a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.minify, minify = _c === void 0 ? true : _c, _d = _b.overrideOptions, overrideOptions = _d === void 0 ? {} : _d;
    return function generateIndexHtml() {
        return {
            plugins: [
                /**
                 * Plugin: HtmlWebpackPlugin
                 * Description: Simplifies creation of HTML files to serve your webpack bundles.
                 * This is especially useful for webpack bundles that include a hash in the filename
                 * which changes every compilation.
                 *
                 * See: https://github.com/ampedandwired/html-webpack-plugin
                 */
                new HtmlWebpackPlugin(Object.assign({
                    template: 'index.html',
                    chunksSortMode: 'dependency',
                    minify: minify ? {
                        removeComments: true,
                        collapseWhitespace: true
                    } : undefined,
                    metadata: core_1.get(this, 'metadata', {})
                }, overrideOptions)),
            ].concat(core_1.get(this, 'plugins', []))
        };
    };
};
